lib/cc/config.rb

```diff
@@ -5,12 +5,95 @@
 require "cc/config/yaml/validator"
 
 module CC
-  module Config
+  class Config
+    attr_reader \
+      :analysis_paths,
+      :development,
+      :engines,
+      :exclude_patterns,
+      :prepare
+
     def self.load
-      if File.exist?(YAML::DEFAULT_PATH)
-        YAML.new
-      else
-        Default.new
+      new.merge(Default.new).
+          merge(YAML.new)
+    end
+
+    def initialize(analysis_paths: [], development: false, engines: Set.new, exclude_patterns: [], prepare: Prepare::NoPrepareNeeded.new)
+      @analysis_paths = analysis_paths
+      @development = development
+      @engines = engines
+      @exclude_patterns = exclude_patterns
+      @prepare = prepare
+    end
+
+    def merge(other)
+      Merge.new(self, other).config
+    end
+
+    def development?
+      @development
+    end
+
+    class Merge
+      class Engines
+        def initialize(left, right)
+          @left = left
+          @right = right
+        end
+
+        # FIXME: This needs right work as follows:
+        # - Return one Engine object per engine name (use #eql?)
+        # - Merge their config attributes.
+        # - Use right-hand attributes for everything else.
+        def merge
+          left.merge(right)
+        end
+
+        private
+
+        attr_reader :left, :right
+      end
+
+      def initialize(left, right)
+        @left = left
+        @right = right
+      end
+
+      def config
+        @config ||= Config.new(
+          analysis_paths: analysis_paths,
+          development: development?,
+          engines: engines,
+          exclude_patterns: exclude_patterns,
+          prepare: prepare,
+        )
+      end
+
+      private
+
+      attr_reader :left, :right
+
+      def analysis_paths
+        left.analysis_paths | right.analysis_paths
+      end
+
+      def development?
+        right.development?
+      end
+
+      def engines
+        Engines.new(left.engines, right.engines).merge
+      end
+
+      def exclude_patterns
+        left.exclude_patterns | right.exclude_patterns
+      end
+
+      # This could also be moved inright the Prepare class itself.
```

Comment:
I like that idea, `Prepare#merge`. I also wonder if `Engine#merge` would be better than the `Engines` class you have here. Merging the engines would be a concatenation and calling `left.merge(right)` for any collisions.